### PR TITLE
PKCS11: Support for Key Generation (Generic Secret and AES Keys)

### PR DESCRIPTION
**Pull request by ruchi393**

### CODE REVIEW

**Review by etienne-lms — COMMENTED** [review]
LGTM

File: libckteec/src/pkcs11_processing.c [hunk 1]
```diff
@@ -556,3 +556,81 @@ CK_RV ck_signverify_final(CK_SESSION_HANDLE session,
...
+		if (rv == CKR_OK)
+			rv = CKR_DEVICE_ERROR;
+		goto bail;
+	}
+
+	memcpy(&key_handle, out_shm->buffer, sizeof(key_handle));
+	*handle = key_handle;
+
+bail:
+	ckteec_free_shm(ctrl);
```
Comment (vesajaaskelainen) [inline libckteec/src/pkcs11_processing.c:630]: Please do free's in reverse order as what they were allocated.

Comment (ruchi393) [reply libckteec/src/pkcs11_processing.c]: Done

**Comment by vesajaaskelainen** [comment]
This https://github.com/OP-TEE/optee_client/pull/241/commits/8a2209b43c28738921b336e1b3fb5e2873d2dad9 should be squashed into https://github.com/OP-TEE/optee_client/pull/241/commits/fc3d5a47a45ed21ec0d8d2fbd43883596bb5aa7a. And then for that:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

This https://github.com/OP-TEE/optee_client/pull/241/commits/c69fa8d7fa198c55bbb73af07d4a0f605af63ec4 should be squashed into https://github.com/OP-TEE/optee_client/pull/241/commits/2d1a49cc29732ce2803072a854905aa6ad1d43ef. And then for that:
`Reviewed-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>`

Assuming that all checks pass after that.

**Comment by ruchi393** [comment]
Thanks @vesajaaskelainen 

Patches squashed and tags applied

**Comment by jforissier** [comment]
> Assuming that all checks pass after that.

IBART doesn't seem to be working right now, I tried restarting it but received a 500 Internal error.

**Comment by jbech-linaro** [comment]
> IBART doesn't seem to be working right now, I tried restarting the job but received a 500 Internal error.

I'm checking ... I'll keep you posted. Maybe "he's" only Christmas holidays :wink:

**Comment by jbech-linaro** [comment]
It's not obvious what went wrong...

``` 
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]: Exception in thread Thread-2254:
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]: Traceback (most recent call last):
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/ibartenv/lib/python3.6/site-packages/pexpect/pty_spawn.py", line 23, in _wrap_ptyprocess_err
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     yield
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/ibartenv/lib/python3.6/site-packages/pexpect/pty_spawn.py", line 327, in close
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     self.ptyproc.close(force=force)
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/ibartenv/lib/python3.6/site-packages/ptyprocess/ptyprocess.py", line 400, in close
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     raise PtyProcessError('Could not terminate the child.')
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]: ptyprocess.util.PtyProcessError: Could not terminate the child.
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]: During handling of the above exception, another exception occurred:
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]: Traceback (most recent call last):
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     self.run()
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/worker.py", line 301, in run
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     current_status = status.d[self.start_job()]
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/worker.py", line 259, in start_job
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     terminate_child(child)
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:   File "/home/optee/devel/ibart/worker.py", line 161, in terminate_child
Dec 23 08:09:04 jbech-E6430s gunicorn[29199]:     child.close()
```

So I restarted the service, but still it didn't work. Since I've been updating lots of packages the last week at the IBART machine and it said that a reboot is needed I thought I'd do that as well.

```
root@jbech-E6430s:~/devel/ibart/logs
# reboot now
Connection to optee.mooo.com closed by remote host.
Connection to optee.mooo.com closed.
jbech@goby ~ 
$ optee
ssh: connect to host optee.mooo.com port 2222: No route to host
```
I.e., it looks like I need to go to the office ... I'll try to go there later today, but if not, then IBART will be offline until the 27th at least.

**Comment by jbech-linaro** [comment]
Just back from the office, it looked there were general issues with the router and/or the ISP. After rebooting the router at the office things seems to be working again. I've restarted this job. We should have the test results shortly.

**Comment by jforissier** [comment]
@jbech-linaro great, thanks!

**Pull request merged**